modules/images/webp-uploads/load.php

```diff
@@ -5,40 +5,294 @@
+ * @since n.e.x.t
  *
- * @since 1.0.0
+ * @see   wp_generate_attachment_metadata
+ * @see   webp_uploads_valid_image_mime_types
  *
- * @param string $output_format The image editor default output format mapping.
- * @param string $filename      Path to the image.
- * @param string $mime_type     The source image mime type.
- * @return string The new output format mapping.
+ * @param array $metadata      An array with the metadata from this attachment.
+ * @param int   $attachment_id The ID of the attachment where the hook was dispatched.
+ *
+ * @return array An array with the updated structure for the metadata before is stored in the database.
  */
-function webp_uploads_filter_image_editor_output_format( $output_format, $filename, $mime_type ) {
-	// Only enable if the server supports WebP.
-	if ( ! wp_image_editor_supports( array( 'mime_type' => 'image/webp' ) ) ) {
-		return $output_format;
+function webp_uploads_create_sources_property( array $metadata, $attachment_id ) {
+	// This should take place only on the JPEG image.
+	$valid_mime_transforms = webp_uploads_get_supported_image_mime_transforms();
+
+	// Not a supported mime type to create the sources property.
+	if ( ! array_key_exists( get_post_mime_type( $attachment_id ), $valid_mime_transforms ) ) {
+		return $metadata;
+	}
+
+	$file = get_attached_file( $attachment_id, true );
+
+	// File does not exist.
+	if ( ! file_exists( $file ) ) {
+		return $metadata;
+	}
+
+	$dirname     = pathinfo( $file, PATHINFO_DIRNAME );
+	$image_sizes = array();
+	if ( array_key_exists( 'sizes', $metadata ) && is_array( $metadata['sizes'] ) ) {
+		$image_sizes = $metadata['sizes'];
+	}
+
+	$sizes = array();
+
+	foreach ( wp_get_registered_image_subsizes() as $size_name => $properties ) {
+		// This image size does not exist on the defined sizes.
+		if ( ! array_key_exists( $size_name, $image_sizes ) || ! is_array( $image_sizes[ $size_name ] ) ) {
+			continue;
+		}
+
+		$current_size = $image_sizes[ $size_name ];
+		$sources      = array();
+		if ( array_key_exists( 'sources', $current_size ) && is_array( $current_size['sources'] ) ) {
+			$sources = $current_size['sources'];
+		}
+
+		// Try to find the mime type of the image size.
+		$current_mime = '';
+		if ( array_key_exists( 'mime-type', $current_size ) ) {
+			$current_mime = $current_size['mime-type'];
+		} elseif ( array_key_exists( 'file', $current_size ) ) {
+			$current_mime = wp_check_filetype( $current_size['file'] )['type'];
+		}
+
+		if ( empty( $current_mime ) ) {
+			$sizes[ $size_name ] = $current_size;
+			continue;
+		}
+
+		$sources[ $current_mime ] = array(
+			'file'     => array_key_exists( 'file', $current_size ) ? $current_size['file'] : '',
+			'filesize' => 0,
+		);
+
+		// Set the filesize from the current mime image.
+		$file_location = path_join( $dirname, $sources[ $current_mime ]['file'] );
+		if ( file_exists( $file_location ) ) {
+			$sources[ $current_mime ]['filesize'] = filesize( $file_location );
+		}
```

Comment:
This whole block here should be wrapped in a condition, we only need to run it `if ( ! isset( $sources[ $current_mime ] )`.

Otherwise it will run this logic over and over again although the data is already set.

Reply:
Right this is actually helpful as if this run when the image is updated we ensure that the `sources` property points to the right image instead of the older one, as this won't be updated in that scenario.

Subsequent updates from:

- #158 

Would handle additional cases like _should previous sources be stored in the backup sizes or delete those.?_

Reply:
Let's handle that part once we actually get to #158 - at this point we don't support editing images, so it's better to focus on exclusively the goal of #142. Right now, executing this logic on every call of this function is unnecessary - we only need to populate the `$sources` value for the existing (primary) image format once, only _if_ it doesn't exist yet.

#158 will show if this is really needed or not - probably the implementation there will have to use different hooks or behavior anyway.

Reply:
Sounds good this has been introduced as an `empty` check inside of the `foreach` loop.